subprojects/base-services/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationExecutorParallelExecutionTest.groovy

```diff
@@ -387,4 +398,44 @@ class DefaultBuildOperationExecutorParallelExecutionTest extends ConcurrentSpec
         operationState != null
         !operationState.running
     }
+
+    def "registers/deregisters a listener for parallelism changes"() {
+        def parallelExecutionManager = parallelism(1)
+
+        when:
+        buildOperationExecutor = new DefaultBuildOperationExecutor(operationListener, Mock(TimeProvider), new NoOpProgressLoggerFactory(),
+            Stub(BuildOperationQueueFactory), Stub(ExecutorFactory), parallelExecutionManager)
+
+        then:
+        1 * parallelExecutionManager.addListener(_)
+
+        when:
+        buildOperationExecutor.stop()
+
+        then:
+        1 * parallelExecutionManager.removeListener(_)
+    }
+
+    def "adjusts thread pool size when parallelism configuration changes"() {
```

Comment:
❌  Need some coverage for **reducing** the pool size while more than targeted size actions are running.